File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/CardFilterInteractorImplTest.java
package com.dbottillo.mtgsearchfree.interactors;

import com.dbottillo.mtgsearchfree.model.CardFilter;
import com.dbottillo.mtgsearchfree.model.storage.CardsPreferencesImpl;
import com.dbottillo.mtgsearchfree.util.Logger;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnit;
import org.mockito.junit.MockitoRule;

import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class CardFilterInteractorImplTest {

    @Rule
    public MockitoRule mockitoRule = MockitoJUnit.rule();

    @Mock
    CardFilter cardFilter;

    @Mock
    CardsPreferencesImpl cardsPreferences;

    @Mock
    Logger logger;

    private CardFilterInteractor underTest;

    @Before
    public void setUp() throws Exception {
        when(cardsPreferences.load()).thenReturn(cardFilter);
        underTest = new CardFilterInteractorImpl(cardsPreferences, logger);
    }

    @Test
    public void willLoadDataFromStorage() {
        TestObserver<CardFilter> testSubscriber = new TestObserver<>();
        underTest.load().subscribe(testSubscriber);

        testSubscriber.assertNoErrors();
        testSubscriber.assertValue(cardFilter);
        verify(cardsPreferences).load();
    }

    @Test
    public void willSyncDataWithStorage() {
        underTest.sync(cardFilter);

        verify(cardsPreferences).sync(cardFilter);
    }

}

File: MTGSearch/src/test/java/com/dbottillo/mtgsearchfree/interactors/CardFilterInteractorImplTest.kt
package com.dbottillo.mtgsearchfree.interactors

import com.dbottillo.mtgsearchfree.model.CardFilter
import com.dbottillo.mtgsearchfree.model.storage.CardsPreferencesImpl
import com.dbottillo.mtgsearchfree.util.Logger

import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.Mock
import org.mockito.junit.MockitoJUnit
import org.mockito.junit.MockitoRule

import io.reactivex.observers.TestObserver
import org.mockito.Mockito.*

class CardFilterInteractorImplTest {

    @Rule @JvmField
    var mockitoRule = MockitoJUnit.rule()

    @Mock
    lateinit var cardFilter: CardFilter

    @Mock
    lateinit var cardsPreferences: CardsPreferencesImpl

    @Mock
    lateinit var logger: Logger

    lateinit var underTest: CardFilterInteractor

    @Before
    fun setUp() {
        underTest = CardFilterInteractorImpl(cardsPreferences, logger)
    }

    @Test
    fun willLoadDataFromStorage() {
        `when`(cardsPreferences.load()).thenReturn(cardFilter)
        val testSubscriber = TestObserver<CardFilter>()

        underTest.load().subscribe(testSubscriber)

        testSubscriber.assertNoErrors()
        testSubscriber.assertValue(cardFilter)
        verify(cardsPreferences).load()
        verifyNoMoreInteractions(cardsPreferences)
    }

    @Test
    fun willSyncDataWithStorage() {
        underTest.sync(cardFilter)

        verify(cardsPreferences).sync(cardFilter)
        verifyNoMoreInteractions(cardsPreferences)
    }

}